ratelimits/gcra_test.go

```diff
@@ -0,0 +1,180 @@
+package ratelimits
+
+import (
+	"testing"
+	"time"
+
+	"github.com/jmhodges/clock"
+	"github.com/letsencrypt/boulder/config"
+	"github.com/letsencrypt/boulder/test"
+)
+
+func Test_decide(t *testing.T) {
+	clk := clock.NewFake()
+	limit := precomputeLimit(
+		limit{Burst: 10, Count: 1, Period: config.Duration{Duration: time.Second}},
+	)
+
+	// Begin by using 1 of our 10 requests.
+	d := maybeSpend(clk, limit, clk.Now(), 1)
+	test.Assert(t, d.Allowed, "should be allowed")
+	test.AssertEquals(t, d.Remaining, 9)
+	test.AssertEquals(t, d.RetryIn, time.Duration(0))
+	test.AssertEquals(t, d.ResetIn, time.Second)
+
+	// Immediately use another 9 of our remaining requests.
+	d = maybeSpend(clk, limit, d.newTAT, 9)
+	test.Assert(t, d.Allowed, "should be allowed")
+	test.AssertEquals(t, d.Remaining, 0)
+	test.AssertEquals(t, d.RetryIn, time.Second)
+	test.AssertEquals(t, d.ResetIn, time.Second*10)
+
+	// Our new TAT should be 10 seconds (limit.Burst) in the future.
+	test.AssertEquals(t, d.newTAT, clk.Now().Add(time.Second*10))
+
+	// Let's try using just 1 more request without waiting.
+	d = maybeSpend(clk, limit, d.newTAT, 1)
+	test.Assert(t, !d.Allowed, "should not be allowed")
+	test.AssertEquals(t, d.Remaining, 0)
+	test.AssertEquals(t, d.RetryIn, time.Second)
+	test.AssertEquals(t, d.ResetIn, time.Second*10)
+
+	// Let's try being exactly as patient as we're told to be.
+	clk.Add(d.RetryIn)
+	d = maybeSpend(clk, limit, d.newTAT, 0)
+	test.AssertEquals(t, d.Remaining, 1)
+
+	// We are 1 second in the future, we should have 1 new request.
+	d = maybeSpend(clk, limit, d.newTAT, 1)
+	test.Assert(t, d.Allowed, "should be allowed")
+	test.AssertEquals(t, d.Remaining, 0)
+	test.AssertEquals(t, d.RetryIn, time.Second)
+	test.AssertEquals(t, d.ResetIn, time.Second*10)
+
+	// Let's try waiting (10 seconds) for our whole bucket to refill.
+	clk.Add(d.ResetIn)
+
+	// We should have 10 new requests. If we use 1 we should have 9 remaining.
+	d = maybeSpend(clk, limit, d.newTAT, 1)
+	test.Assert(t, d.Allowed, "should be allowed")
+	test.AssertEquals(t, d.Remaining, 9)
+	test.AssertEquals(t, d.RetryIn, time.Duration(0))
+	test.AssertEquals(t, d.ResetIn, time.Second)
+
+	// Have you ever tried spending 0, like, just to see what happens?
+	d = maybeSpend(clk, limit, d.newTAT, 0)
+	test.Assert(t, d.Allowed, "should be allowed")
+	test.AssertEquals(t, d.Remaining, 9)
+	test.AssertEquals(t, d.RetryIn, time.Duration(0))
+	test.AssertEquals(t, d.ResetIn, time.Second)
+
+	// Spending 0 simply informed us that we still have 9 remaining, let's see
+	// what we have after waiting 20 hours.
+	clk.Add(20 * time.Hour)
+
+	// C'mon, big money, no whammies, no whammies, STOP!
+	d = maybeSpend(clk, limit, d.newTAT, 0)
```

Comment:
Being allowed to spend 0 is not the point of the test, it's just a way for us to check our token balance, which we go on to do on line `78`. We're checking our balance to ensure that, even after waiting 20 hours, we only accrued 1 additional token because our bucket filled within the first second.